src/types/graphs/viewgraph.ts

```diff
@@ -223,6 +224,80 @@ export class ViewGraph {
     return Array.from(edges).map(([, edge]) => edge);
   }
 
+  getVisibleConnectedDeviceIds(deviceId: DeviceId): DeviceId[] {
+    const visited = new Set<DeviceId>(); // Tracks visited devices to avoid cycles
+    const visibleDevices: DeviceId[] = []; // Stores visible connected device IDs
+
+    const dfs = (currentId: DeviceId): void => {
+      if (visited.has(currentId)) {
+        return; // Skip if the device has already been visited
+      }
+      visited.add(currentId); // Mark the current device as visited
+
+      const currentDevice = this.graph.getVertex(currentId);
+      if (!currentDevice) {
+        return; // Skip if the device does not exist in the graph
+      }
+
+      // If the device is visible and not the starting device, add it to the result
+      if (currentDevice.visible && currentId !== deviceId) {
+        visibleDevices.push(currentId);
+        return; // Stop further traversal from this device
+      }
+
+      // Traverse neighbors of the current device
+      const neighbors = this.graph.getNeighbors(currentId);
+      for (const neighborId of neighbors) {
+        dfs(neighborId);
+      }
+    };
+
+    // Start DFS for each visible neighbor of the given device
+    const neighbors = this.graph.getNeighbors(deviceId);
+    for (const neighborId of neighbors) {
+      dfs(neighborId);
+    }
+
+    return visibleDevices; // Return the list of visible connected device IDs
+  }
+
+  canReachVisibleDevice(
+    startId: DeviceId,
+    excludeId?: DeviceId,
+  ): Set<DeviceId> {
+    const visited = new Set<DeviceId>();
+    const visibleDevices = new Set<DeviceId>();
+
+    const dfs = (currentId: DeviceId): void => {
```

Comment:
We could expose this as a function of `Graph`, since it's a standard DFS with a filter